File: core/src/agent/web/ui/src/lib/activities.ts
import { api } from "@/lib/api";

export const ACTIVITY_STATES = ["active", "paused", "completed", "cancelled"] as const;
export type ActivityState = (typeof ACTIVITY_STATES)[number];
export type ActivityResource = { label: string; reference: string };
export type ActivityDraft = {
  title: string;
  goal: string;
  completion_criteria: string;
  boundaries: string;
  resources: ActivityResource[];
};
export type Activity = ActivityDraft & {
  id: string;
  owner_uid: number;
  state: ActivityState;
  completion_note: string | null;
  created_at: string;
  updated_at: string;
};
export type JobStatus = "pending" | "running" | "waiting_approval" | "ok" | "error" | "cancelled";
export type ActivityJob = {
  id: string;
  title: string;
  status: JobStatus;
  session_id: string | null;
  created_at: string;
  finished_at: string | null;
  response: string | null;
  error: string | null;
  waiting_on: string[];
};
export type ActivityDetail = {
  schema: 1;
  activity: Activity;
  jobs: ActivityJob[];
  sessions: string[];
};
export type ActivityRun = {
  prompt?: string;
  session_id?: string;
  max_turns?: number;
  use_memory?: boolean;
};
export type AppObjectRef = {
  app_id: string;
  object_type: string;
  object_id: string;
  revision?: string;
};
export type ActivityObjectAttachment = { label: string; object: AppObjectRef };
export type ObjectDescription = {
  object: AppObjectRef;
  reference: string;
  app_name: string;
  app_version: string;
  object_label: string;
  object_summary: string;
  invocation: { app_id: string; operation: string; args: string[] };
};
export type ActivityObject = { label: string; reference: string } & (
  | { status: "declared"; description: ObjectDescription; error: null }
  | { status: "unavailable" | "invalid"; description: null; error: string }
);
export type ActivityObjects = {
  schema: 1;
  activity_id: string;
  objects: ActivityObject[];
};
type SubmittedJob = {
  id: string;
  status: JobStatus;
  activity_id: string;
  session_id?: string | null;
};

export const activityStateLabels: Record<ActivityState, string> = {
  active: "Active",
  paused: "Paused",
  completed: "Completed",
  cancelled: "Cancelled",
};
export const jobStatusLabels: Record<JobStatus, string> = {
  pending: "Queued",
  running: "Running",
  waiting_approval: "Waiting for approval",
  ok: "Succeeded",
  error: "Failed",
  cancelled: "Cancelled",
};

export function hasLiveActivityJobs(jobs: ActivityJob[]): boolean {
  return jobs.some(({ status }) =>
    status === "pending" || status === "running" || status === "waiting_approval",
  );
}

function record(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function nullableString(value: unknown): value is string | null {
  return value === null || typeof value === "string";
}

function strings(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string");
}

export function isActivityState(value: unknown): value is ActivityState {
  return ACTIVITY_STATES.some((state) => state === value);
}

function isJobStatus(value: unknown): value is JobStatus {
  return value === "pending" || value === "running" || value === "waiting_approval"
    || value === "ok" || value === "error" || value === "cancelled";
}

function isActivity(value: unknown): value is Activity {
  return record(value)
    && typeof value.id === "string"
    && typeof value.owner_uid === "number"
    && typeof value.title === "string"
    && typeof value.goal === "string"
    && typeof value.completion_criteria === "string"
    && typeof value.boundaries === "string"
    && Array.isArray(value.resources)
    && value.resources.every((item) =>
      record(item) && typeof item.label === "string" && typeof item.reference === "string",
    )
    && isActivityState(value.state)
    && nullableString(value.completion_note)
    && typeof value.created_at === "string"
    && typeof value.updated_at === "string";
}

function isActivityJob(value: unknown): value is ActivityJob {
  return record(value)
    && typeof value.id === "string"
    && typeof value.title === "string"
    && isJobStatus(value.status)
    && nullableString(value.session_id)
    && typeof value.created_at === "string"
    && nullableString(value.finished_at)
    && nullableString(value.response)
    && nullableString(value.error)
    && strings(value.waiting_on);
}

export function readActivity(value: unknown): Activity {
  if (!isActivity(value)) throw new Error("Invalid Activity response from the server.");
  return value;
}

export function readActivityList(value: unknown): Activity[] {
  if (!record(value) || value.schema !== 1 || !Array.isArray(value.activities)
      || !value.activities.every(isActivity)) {
    throw new Error("Invalid Activity list from the server.");
  }
  return value.activities;
}

export function readActivityDetail(value: unknown, id: string): ActivityDetail {
  if (!record(value) || value.schema !== 1 || !isActivity(value.activity)
      || value.activity.id !== id || !Array.isArray(value.jobs)
      || !value.jobs.every(isActivityJob) || !strings(value.sessions)) {
    throw new Error("Invalid Activity detail from the server.");
  }
  return { schema: 1, activity: value.activity, jobs: value.jobs, sessions: value.sessions };
}

function isObjectDescription(value: unknown): value is ObjectDescription {
  return record(value) && record(value.object)
    && typeof value.object.app_id === "string"
    && typeof value.object.object_type === "string"
    && typeof value.object.object_id === "string"
    && (value.object.revision === undefined || typeof value.object.revision === "string")
    && typeof value.reference === "string"
    && typeof value.app_name === "string"
    && typeof value.app_version === "string"
    && typeof value.object_label === "string"
    && typeof value.object_summary === "string"
    && record(value.invocation)
    && value.invocation.app_id === value.object.app_id
    && typeof value.invocation.operation === "string"
    && strings(value.invocation.args);
}

function isActivityObject(value: unknown): value is ActivityObject {
  if (!record(value) || typeof value.label !== "string" || typeof value.reference !== "string") {
    return false;
  }
  if (value.status === "declared") {
    return value.error === null && isObjectDescription(value.description)
      && value.description.reference === value.reference;
  }
  return (value.status === "unavailable" || value.status === "invalid")
    && value.description === null && typeof value.error === "string";
}

export function readActivityObjects(value: unknown, id: string): ActivityObjects {
  if (!record(value) || value.schema !== 1 || value.activity_id !== id
      || !Array.isArray(value.objects) || !value.objects.every(isActivityObject)) {
    throw new Error("Invalid App object descriptions from the server.");
  }
  return { schema: 1, activity_id: id, objects: value.objects };
}

function readSubmittedJob(value: unknown, id: string): SubmittedJob {
  if (!record(value) || typeof value.id !== "string" || !isJobStatus(value.status)
      || value.activity_id !== id
      || (value.session_id !== undefined && !nullableString(value.session_id))) {
    throw new Error("Invalid submitted Activity job from the server. Refresh before retrying.");
  }
  return {
    id: value.id,
    status: value.status,
    activity_id: id,
    session_id: value.session_id,
  };
}

const activityPath = (id: string) => `/api/activities/${encodeURIComponent(id)}`;

export const activityApi = {
  list: async (state?: ActivityState, signal?: AbortSignal) => readActivityList(
    await api.get<unknown>(`/api/activities?limit=100${state ? `&state=${state}` : ""}`, { signal }),
  ),
  get: async (id: string, signal?: AbortSignal) => readActivityDetail(
    await api.get<unknown>(`${activityPath(id)}?limit=100`, { signal }), id,
  ),
  create: async (draft: ActivityDraft) => readActivity(
    await api.post<unknown>("/api/activities", draft),
  ),
  update: async (id: string, draft: ActivityDraft) => readActivity(
    await api.post<unknown>(`${activityPath(id)}/update`, draft),
  ),
  transition: async (id: string, state: ActivityState, completionNote?: string) => readActivity(
    await api.post<unknown>(`${activityPath(id)}/transition`, {
      state,
      ...(completionNote === undefined ? {} : { completion_note: completionNote }),
    }),
  ),
  run: async (id: string, input: ActivityRun) => readSubmittedJob(
    await api.post<unknown>(`${activityPath(id)}/run`, input), id,
  ),
  objects: async (id: string, signal?: AbortSignal) => readActivityObjects(
    await api.get<unknown>(`${activityPath(id)}/objects`, { signal }), id,
  ),
  attachObject: async (id: string, attachment: ActivityObjectAttachment) => {
    const activity = readActivity(await api.post<unknown>(`${activityPath(id)}/objects`, attachment));
    if (activity.id !== id) throw new Error("Invalid attached Activity from the server. Refresh before retrying.");
    return activity;
  },
};

File: desktop/agent/protocol/src/activities.rs
//! Activity presentation only. Ownership, persistence, transitions and work
//! admission belong to clawd, shared with the terminal client.

use serde::{Deserialize, Serialize};

use crate::CapabilityPolicyScope;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityState {
    Active,
    Paused,
    Completed,
    Cancelled,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityResource {
    pub label: String,
    pub reference: String,
}

/// Components only; canonical URI semantics belong to the shared broker/SDK.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct AppObjectReference {
    pub app_id: String,
    pub object_type: String,
    pub object_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityObjectAttachRequest {
    pub label: String,
    pub object: AppObjectReference,
}

/// A declaration is authenticated metadata, not proof of object access.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityObjectStatus {
    Declared,
    Unavailable,
    Invalid,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AppObjectInvocation {
    pub app_id: String,
    pub operation: String,
    #[serde(default)]
    pub args: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityOperationPreviewRequest {
    pub app_id: String,
    pub operation: String,
    pub args: Vec<String>,
}

impl From<&AppObjectInvocation> for ActivityOperationPreviewRequest {
    fn from(invocation: &AppObjectInvocation) -> Self {
        Self {
            app_id: invocation.app_id.clone(),
            operation: invocation.operation.clone(),
            args: invocation.args.clone(),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppEffectKind {
    Read,
    Create,
    Update,
    Delete,
    External,
    Execute,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppEffectRecovery {
    NotApplicable,
    Reversible,
    Compensatable,
    Irreversible,
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppEffectTargetKind {
    Path,
    Host,
    Name,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppEffectTargetState {
    Requested,
    Unspecified,
    Unresolved,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AppDeclaredEffect {
    pub kind: AppEffectKind,
    pub label: String,
    pub recovery: AppEffectRecovery,
    pub target_arg: Option<String>,
    pub target_kind: Option<AppEffectTargetKind>,
    pub requested_targets: Vec<String>,
    pub target_state: AppEffectTargetState,
}

/// Manifest metadata only, never execution evidence or an authorization grant.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityOperationPreview {
    pub schema: u32,
    pub app_id: String,
    pub app_name: String,
    pub app_version: String,
    pub package_digest: String,
    pub operation: String,
    pub operation_label: String,
    pub effects_declared: bool,
    pub effects: Vec<AppDeclaredEffect>,
    pub unresolved_arguments: Vec<String>,
    pub authorization_checked: bool,
    pub executed: bool,
    pub effects_confirmed: bool,
    pub notes: Vec<String>,
}

impl ActivityOperationPreview {
    pub fn is_metadata_only(&self) -> bool {
        self.schema == 1 && !self.authorization_checked && !self.executed && !self.effects_confirmed
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityReceiptSource {
    CallerReported,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityReceiptOutcome {
    Returned,
    ReportedError,
    Indeterminate,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReceiptResultKind {
    Json,
    Text,
    Empty,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReceiptResultSummary {
    pub kind: ReceiptResultKind,
    pub sha256: String,
    pub bytes: u64,
    pub preview: String,
    pub preview_truncated: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityReceiptReport {
    pub id: String,
    pub app_id: String,
    pub operation: String,
    pub package_digest: String,
    pub outcome: ActivityReceiptOutcome,
    #[serde(default)]
    pub result: Option<ReceiptResultSummary>,
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReceiptDeclaredEffect {
    pub kind: AppEffectKind,
    pub label: String,
    pub recovery: AppEffectRecovery,
    #[serde(default)]
    pub target_arg: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityReceiptDeclaration {
    pub app_version: String,
    pub operation_label: String,
    pub effects: Vec<ReceiptDeclaredEffect>,
}

/// An immutable caller report, not an OS execution or mutation attestation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityReceiptView {
    pub id: String,
    pub activity_id: String,
    pub received_at: String,
    pub source: ActivityReceiptSource,
    pub report: ActivityReceiptReport,
    #[serde(default)]
    pub declaration: Option<ActivityReceiptDeclaration>,
    #[serde(default)]
    pub declaration_error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityReceiptsResponse {
    pub schema: u32,
    pub activity_id: String,
    pub receipts: Vec<ActivityReceiptView>,
}

impl ActivityReceiptsResponse {
    pub fn matches_activity(&self, id: &str) -> bool {
        self.schema == 1
            && !id.trim().is_empty()
            && self.activity_id == id
            && self
                .receipts
                .iter()
                .all(|receipt| !receipt.id.trim().is_empty() && receipt.activity_id == id)
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityReceiptsQuery {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AppObjectDescription {
    pub object: AppObjectReference,
    pub reference: String,
    #[serde(default)]
    pub app_name: String,
    #[serde(default)]
    pub app_version: String,
    #[serde(default)]
    pub object_label: String,
    #[serde(default)]
    pub object_summary: String,
    pub invocation: AppObjectInvocation,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityObjectResourceView {
    pub label: String,
    pub reference: String,
    pub status: ActivityObjectStatus,
    #[serde(default)]
    pub description: Option<AppObjectDescription>,
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityObjectsResponse {
    pub activity_id: String,
    #[serde(default)]
    pub objects: Vec<ActivityObjectResourceView>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityView {
    pub id: String,
    pub title: String,
    pub goal: String,
    pub state: ActivityState,
    #[serde(default)]
    pub completion_criteria: String,
    #[serde(default)]
    pub boundaries: String,
    #[serde(default)]
    pub resources: Vec<ActivityResource>,
    #[serde(default)]
    pub completion_note: Option<String>,
    #[serde(default)]
    pub created_at: String,
    #[serde(default)]
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityJobView {
    pub id: String,
    pub status: String,
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub created_at: String,
    #[serde(default)]
    pub finished_at: Option<String>,
    #[serde(default)]
    pub response: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub waiting_on: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityApprovalView {
    pub id: String,
    pub session_id: String,
    pub label: String,
    #[serde(default)]
    pub reason: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityDecisionStatus {
    Pending,
    Approved,
    Denied,
    Unavailable,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityIssueKind {
    WaitingApproval,
    Indeterminate,
    Failed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ActivityDecisionRisk {
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ActivityNotificationSeverity {
    Info,
    Warning,
    Error,
    Critical,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityNotificationState {
    Unread,
    Read,
    Acknowledged,
    Dismissed,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionCounts {
    pub queued: u64,
    pub running: u64,
    pub waiting: u64,
    pub completed: u64,
    pub failed: u64,
    pub cancelled: u64,
    pub indeterminate: u64,
    pub pending_decisions: u64,
    pub unavailable_decisions: u64,
    pub unread_notifications: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionDecision {
    pub id: String,
    pub job_id: String,
    pub session_id: Option<String>,
    pub status: ActivityDecisionStatus,
    pub requested_at: Option<u64>,
    pub verb: Option<String>,
    pub scope: Option<CapabilityPolicyScope>,
    pub risk: Option<ActivityDecisionRisk>,
    pub reason: Option<String>,
    pub review_id: Option<String>,
    pub error: Option<String>,
}

impl ActivityAttentionDecision {
    fn is_consistent(&self) -> bool {
        if self.id.is_empty() || self.job_id.is_empty() {
            return false;
        }
        let details = self.requested_at.is_some()
            && self.verb.as_ref().is_some_and(|value| !value.is_empty())
            && self.scope.is_some()
            && self.risk.is_some()
            && self.reason.is_some()
            && self
                .review_id
                .as_ref()
                .is_some_and(|value| !value.is_empty())
            && self.error.is_none();
        let unavailable = self.requested_at.is_none()
            && self.verb.is_none()
            && self.scope.is_none()
            && self.risk.is_none()
            && self.reason.is_none()
            && self.review_id.is_none()
            && self.error.as_ref().is_some_and(|value| !value.is_empty());
        match self.status {
            ActivityDecisionStatus::Unavailable => unavailable,
            _ => details,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionIssue {
    pub job_id: String,
    pub session_id: Option<String>,
    pub kind: ActivityIssueKind,
    pub status: String,
    pub execution_phase: String,
    pub title: String,
    pub created_at: String,
    pub finished_at: Option<String>,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionNotification {
    pub id: String,
    pub source: String,
    pub kind: String,
    pub severity: ActivityNotificationSeverity,
    pub title: String,
    pub body: String,
    pub task_id: Option<String>,
    pub session_id: Option<String>,
    pub state: ActivityNotificationState,
    pub updated_at_ms: i64,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionTotals {
    pub decisions: u64,
    pub issues: u64,
    pub notifications: u64,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionHasMore {
    pub decisions: bool,
    pub issues: bool,
    pub notifications: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityAttentionResponse {
    pub schema: u32,
    pub activity_id: String,
    pub activity_state: ActivityState,
    pub limit: u32,
    pub counts: ActivityAttentionCounts,
    pub decisions: Vec<ActivityAttentionDecision>,
    pub issues: Vec<ActivityAttentionIssue>,
    pub notifications: Vec<ActivityAttentionNotification>,
    pub totals: ActivityAttentionTotals,
    pub has_more: ActivityAttentionHasMore,
}

impl ActivityAttentionResponse {
    pub fn matches_activity(&self, id: &str, state: ActivityState) -> bool {
        self.schema == 1
            && self.activity_id == id
            && self.activity_state == state
            && (1..=100).contains(&self.limit)
            && self.decisions.len() <= self.limit as usize
            && self.issues.len() <= self.limit as usize
            && self.notifications.len() <= self.limit as usize
            && self
                .decisions
                .iter()
                .all(ActivityAttentionDecision::is_consistent)
            && self.issues.iter().all(|issue| {
                !issue.job_id.is_empty()
                    && !issue.status.is_empty()
                    && !issue.execution_phase.is_empty()
            })
            && self.notifications.iter().all(|notification| {
                !notification.id.is_empty()
                    && notification
                        .task_id
                        .as_ref()
                        .is_some_and(|id| !id.is_empty())
            })
            && page_matches(
                self.totals.decisions,
                self.decisions.len(),
                self.has_more.decisions,
            )
            && page_matches(self.totals.issues, self.issues.len(), self.has_more.issues)
            && page_matches(
                self.totals.notifications,
                self.notifications.len(),
                self.has_more.notifications,
            )
    }
}

fn page_matches(total: u64, shown: usize, has_more: bool) -> bool {
    total >= shown as u64 && has_more == (total > shown as u64)
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityListResponse {
    #[serde(default)]
    pub activities: Vec<ActivityView>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityDetailResponse {
    pub activity: ActivityView,
    #[serde(default)]
    pub jobs: Vec<ActivityJobView>,
    #[serde(default)]
    pub sessions: Vec<String>,
    #[serde(default)]
    pub pending_approvals: Vec<ActivityApprovalView>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub approvals_error: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attention: Option<ActivityAttentionResponse>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attention_error: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityListQuery {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub state: Option<ActivityState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityCreateRequest {
    pub title: String,
    pub goal: String,
    #[serde(default)]
    pub completion_criteria: String,
    #[serde(default)]
    pub boundaries: String,
    #[serde(default)]
    pub resources: Vec<ActivityResource>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityUpdateRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub goal: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub completion_criteria: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub boundaries: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resources: Option<Vec<ActivityResource>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityTransitionRequest {
    pub state: ActivityState,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub completion_note: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivityRunRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prompt: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub use_memory: Option<bool>,
}

/// Acknowledges durable work admission, not achievement of an Activity's goal.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivityWorkResponse {
    pub id: String,
    pub status: String,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub activity_id: Option<String>,
}

#[cfg(test)]
mod tests {
    include!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test/unit/activities.rs"
    ));
}
